include/caffe/util/thread_pool.hpp

```diff
@@ -0,0 +1,112 @@
+#ifndef THREAD_POOL_H
```

Comment:
I would suggest to use Boost's thread_group (essentially a pool) with boost_asio,
Here is a small example:
http://stackoverflow.com/questions/19500404/how-to-create-a-thread-pool-using-boost-in-c
In our case, we may in fact also use asio for reading.